components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/NetworkMessageEncoder.cs

```diff
@@ -429,30 +434,35 @@ public class NetworkMessageEncoder : IMessageEncoder {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            if (_knownPayloadIdentifiers.TryGetValue(notification.NodeId.ToString(), out var knownPayloadIdentifier)) {
-                if (!string.IsNullOrEmpty(knownPayloadIdentifier)) {
-                    return knownPayloadIdentifier;
-                }
+            if (_knownPayloadIdentifiers.TryGetValue(notification.NodeId.ToString(), out var knownPayloadIdentifier) && !string.IsNullOrEmpty(knownPayloadIdentifier)) {
+                return knownPayloadIdentifier;
             }
-            else { //do the long running lookup as less as possible
-                foreach (var dataSetWriter in message.WriterGroup.DataSetWriters) {
-                    foreach (var publishedVariableData in dataSetWriter.DataSet.DataSetSource.PublishedVariables.PublishedData) {
-                        if ((publishedVariableData.PublishedVariableNodeId == notification.NodeId
-                              || publishedVariableData.PublishedVariableNodeId.ToExpandedNodeId(context).AsString(context) == notification.NodeId.ToExpandedNodeId(context.NamespaceUris).AsString(context)) &&
-                            publishedVariableData.Id != notification.NodeId) {
+            else {
+                //do the long running lookup as less as possible
+                var dataSetWriter = message.Writer;
+                foreach (var publishedVariableData in dataSetWriter.DataSet.DataSetSource.PublishedVariables.PublishedData) {
+                    if (publishedVariableData.PublishedVariableNodeId == notification.NodeId
+                        || publishedVariableData.PublishedVariableNodeId.ToExpandedNodeId(context).AsString(context) == notification.NodeId.ToExpandedNodeId(context.NamespaceUris).AsString(context)) {
```

Comment:
>notification.NodeId.ToExpandedNodeId(context.NamespaceUris).AsString(context) [](http://example.com/codeflow?start=120&length=77)

Avoid multiple calls to notification.NodeId.ToExpandedNodeId(context.NamespaceUris).AsString(context) store it as local variable

Reply:
Fixed.